feat: support remote storage URI injection for serving runtimes

de0725 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. Before raising a PR, please run `make precommit` to check the code style.
3. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
4. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
1. Injects REMOTE_STORAGE_URI= into the main user container.
2. Respects both storageSpec-based and PVC-based source URIs.
3. Adds corresponding unit tests to ensure correctness and backward compatibility.
[REMOTE_STORAGE_URI.Usage.and.Testing.Guide (1).docx](https://github.com/user-attachments/files/20422403/REMOTE_STORAGE_URI.Usage.and.Testing.Guide.1.docx)


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3428 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:
Added unit test: TestRemoteStorageUriEnvInjection
Verified behavior under both ConfigMap and Annotation-based configurations
Ran make test, all unit tests passed

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Inject REMOTE_STORAGE_URI into user container when remote storage is enabled, allowing runtimes to access the original model source URI.
```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

de0725 committed:
feat: support remote storage URI injection for serving runtimes

Signed-off-by: de0725 <debosora.52901@gmail.com>

de0725 commented:
Hi @cmaddalozzo @spolti @yuzisun 
Would you have some time to review this PR ?
Thank you very much for your time and support!

spolti commented on pkg/webhook/admission/pod/storage_initializer_injector.go:
```diff
@@ -868,3 +888,12 @@ func needCaBundleMount(caBundleConfigMapName string, initContainer *corev1.Conta
 	return result
 }
+
+func hasEnv(container *corev1.Container, key string) bool {
```
We have a similar function at `pkg/utils/utils.go` GetEnvVarValue.
If that is not suitable for this use case, can you please move this function to the package mentioned above?

de0725 replied:
Got it. I’ll review and move the function to the suggested package as soon as possible. Thank you for pointing it out!

de0725 replied:
Hi @spolti , I’ve removed the custom hasEnv and replaced it with utils.GetEnvVarValue, added the necessary import. Please take another look and let me know if any further tweaks are needed. Thanks!

spolti commented on pkg/constants/constants.go:
```diff
@@ -69,6 +69,12 @@ var (
 
+// Remote Storage URI
+const (
+	EnableRemoteStorageEnvAnnotation = "kserve.io/enable-remote-storage-env"
```
reading the related issue, adding a flag for it seems not to address the issue.
@cmaddalozzo can you help clarifying it?

de0725 replied:
Hi @spolti and @cmaddalozzo  ,

Just to double-check that I'm on the same page:  
When we overwrite the `STORAGE_URI` environment variable with `/mnt/models`, are we expected to:

1. **Always** inject a new environment variable (e.g., `REMOTE_STORAGE_URI`) with the original URI as a default behavior?

**or**

2. Make this behavior optional and controlled via an annotation flag like `kserve.io/enable-remote-storage-env`?

My initial understanding was option 2, but based on your comments it sounds like option 1 — making it the default behavior — might be preferred. 
Would love to confirm before proceeding with the update.

Thanks again for your guidance!

yuzisun replied:
@de0725 I think we can always inject the `REMOTE_STORAGE_URI` environment variable if storageURI presents on YAML, we do not need a flag to enable this as a new environment variable should not affect anything.

de0725 replied:
Got it, I’ll update the logic to always inject REMOTE_STORAGE_URI and submit a PR soon. Thanks!

yuzisun commented on pkg/webhook/admission/pod/storage_initializer_injector.go:
```diff
@@ -237,6 +251,13 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *corev1.Pod)
 
+	if enabledRemoteEnv {
+		if _, exists := utils.GetEnvVarValue(userContainer.Env, RemoteStorageEnvVarName); exists {
+			return fmt.Errorf("container %s already defines %s", userContainer.Name, RemoteStorageEnvVarName)
```
The webhook can be called multiple times so we need to make it idempotent, we do not need to return an error instead check if the environment does not exist we inject it.

de0725 replied:
Good point, I’ll update it to make the injection idempotent as suggested. Thanks!

de0725 committed:
chore: update tests and unconditionally inject REMOTE_STORAGE_URI env var

Signed-off-by: de0725 <debosora.52901@gmail.com>

de0725 committed:
Merge branch 'master' into feature/storage-uri-revamp

de0725 commented:
Hi @yuzisun @spolti  , I’ve refactored the logic to always inject REMOTE_STORAGE_URI when the storageURI annotation is present, removed all related flag checks, and updated the unit tests accordingly. Please review and let me know if any further adjustments are required. Thanks!

de0725 committed:
Merge branch 'master' into feature/storage-uri-revamp

de0725 committed:
Merge branch 'master' into feature/storage-uri-revamp

de0725 committed:
Merge branch 'master' into feature/storage-uri-revamp

de0725 committed:
chore: apply precommit formatting

Signed-off-by: de0725 <debosora.52901@gmail.com>

de0725 commented:
Hi @spolti  @yuzisun ,
Apologies for missing the pre-commit issues earlier. I've fixed the formatting and force-pushed the changes.
Appreciate your time and review. Thanks!

yuzisun reviewed (approved):
Looks good! Thanks @de0725 for the contribution !
